File: MergeSort/merge_sort_step.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <list>
#include <chrono>

using namespace std;

// Function declarations
void mergeSort(vector<int>& S, vector<string>& id, int left, int right);
void merge(vector<int>& S, vector<string>& id, int left, int mid, int right);
void writeFile(const vector<int>& S, const vector<string>& id);

static int startrow = 1;
static int endrow = 7;
static string output = "merge_sort_step_"+ to_string(startrow) + "_" + to_string(endrow) + ".txt";

int main() {
    vector<int> S;
    vector<string> idArray;
    string filePath = "MergeSort/testDataSet.csv"; // Adjust path if needed
    ifstream file(filePath);
    string line;
    int currentrow = 1;

    if (!file.is_open()) {
        cerr << "Could not open file: " << filePath << endl;
        return 1;
    }

    while (getline(file, line)) {
        stringstream ss(line);
        string numStr, idStr;
        if (currentrow >= startrow && currentrow <= endrow){
            if (getline(ss, numStr, ',') && getline(ss, idStr)) {
                S.push_back(stoi(numStr));
                idArray.push_back(idStr);
            }
        }
        currentrow++;
    }

    file.close();

    ofstream clearFile(output, ios::out);
    if (!clearFile) {
        cerr << "Error clearing file." << endl;
        return 0;
    }
    clearFile.close();

    writeFile(S, idArray);
    mergeSort(S, idArray, 0, S.size() - 1);


    return 0;
}

void mergeSort(vector<int>& S, vector<string>& id, int left, int right) {
    if (left >= right) return;

    int mid = (left + right) / 2;
    mergeSort(S, id, left, mid);
    mergeSort(S, id, mid + 1, right);
    merge(S, id, left, mid, right);
}

void merge(vector<int>& S, vector<string>& id, int left, int mid, int right) {
    vector<int> L, R;
    vector<string> Lid, Rid;

    for (int i = left; i <= mid; ++i) {
        L.push_back(S[i]);
        Lid.push_back(id[i]);
    }
    for (int j = mid + 1; j <= right; ++j) {
        R.push_back(S[j]);
        Rid.push_back(id[j]);
    }

    int n1 = mid - left + 1;
    int n2 = right - mid;
    int i = 0, j = 0, k = left;

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            S[k] = L[i];
            id[k] = Lid[i];
            i++;
        } else {
            S[k] = R[j];
            id[k] = Rid[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        S[k] = L[i];
        id[k] = Lid[i];
        i++; k++;
    }

    while (j < n2) {
        S[k] = R[j];
        id[k] = Rid[j];
        j++; k++;
    }
    writeFile(S, id);
}

void writeFile(const vector<int>& S, const vector<string>& id) {
    // Write the opening bracket
    ofstream writer(output, ios::app);
    if (!writer) {
        cerr << "Error opening file: " << output << endl;
        return;
    }
    writer << "[";

    // Write each num/id pair
    for (size_t i = 0; i < S.size(); ++i) {
        writer << S[i] << "/" << id[i] << ", ";
    }

    // Write the closing bracket and newline
    writer << "]\n";
    writer.close();
}

File: MergeSort/merge_sort_step.java
package MergeSort;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedList;

public class merge_sort_step {
    static int startrow = 1;
    static int endrow = 7;
    static String output = "merge_sort_step_"+ startrow +"_" + endrow + ".txt";
  public static void main (String args[]) { 
    ArrayList<Integer> S = new ArrayList<>();
    ArrayList<String> idArray = new ArrayList<>();
    String line;
    String filePath = "MergeSort/testDataSet.csv"; // Path to your CSV file
    //String filePath = "Assingment/dataset_sample_1000.csv";
    int currentrow = 1;

    try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
        while ((line = br.readLine()) != null) {
            // Split the line by comma
            if (currentrow >= startrow && currentrow <= endrow){
                String[] values = line.split(",");
                S.add(Integer.parseInt(values[0]));
                idArray.add(values[1]);
            }
            currentrow++;
            
        }
    } catch (IOException e) {
        e.printStackTrace();
    }
    //Overrides outputfile
    try (BufferedWriter writer = new BufferedWriter(new FileWriter(output))) {
            writer.write("");
        } catch (IOException e) {
            e.printStackTrace();
        }
    writeFile(S, idArray);

    mergeSort (S, idArray); 


  } 

  static void mergeSort (ArrayList<Integer> S, ArrayList<String> id) { 

    mergeSort (S, id, 0, S.size()-1); 

  } 


  static void mergeSort (ArrayList<Integer> S, ArrayList<String> id, int left, int right) { 

    // WRITE YOUR CODE 
    int mid = (left+right)/2;
    if (left != right){
        mergeSort (S, id, left, mid);
        mergeSort (S, id, mid + 1, right);
        merge (S, id, left, mid, right);
    }
    
  } 

  static void merge (ArrayList<Integer> S, ArrayList<String> id, int left, int mid, int right) { 

    // Create LinkedList L from S[left]...S[mid] 

    LinkedList<Integer> L = new LinkedList<>(); 
    LinkedList<String> Lid = new LinkedList<>(); 

    for (int i = 0; i < mid - left + 1; i++) 
    {
        L.add (S.get(left + i)); 
        Lid.add(id.get(left + i)); 
    }
 

    // Create LinkedList R from S[mid+1]...S[right] 

    LinkedList<Integer> R = new LinkedList<>(); 
    LinkedList<String> Rid = new LinkedList<>();

    for (int j = 0; j < right - mid; j++) 
    {
        R.add (S.get(mid + 1 + j)); 
        Rid.add(id.get(mid + 1 + j)); 
    }

    int k = left; 

    while (!L.isEmpty() &&  !R.isEmpty()){
        if (L.getFirst() < R.getFirst()){
            S.set(k, L.removeFirst());
            id.set(k, Lid.removeFirst());
            k++;
        }else{
            S.set(k, R.removeFirst());
            id.set(k, Rid.removeFirst());
            k++;
        }

    }
  
    while (!L.isEmpty()){
        S.set(k, L.removeFirst());
        id.set(k, Lid.removeFirst());
        k++;
    } 
  
    while (!R.isEmpty()){
        S.set(k, R.removeFirst());
        id.set(k, Rid.removeFirst());
        k++;
    } 
    writeFile(S, id);
  } 
  public static void writeFile(ArrayList<Integer> num, ArrayList<String> id){

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(output,true))) {
            writer.write("[");
        } catch (IOException e) {
            e.printStackTrace();
        }
        for(int i = 0; i < num.size(); i++){
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(output, true))) {
        
            writer.write(num.get(i) + "/" + id.get(i) + ", ");
            //writer.newLine();  // Adds a new line
            }
            catch (IOException e) {
                e.printStackTrace();
            }
        }
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(output, true))) {
            writer.write("]\n");
            }
            catch (IOException e) {
                e.printStackTrace();
            }
    }
    

}
